File: python/LinkedList/Scaler/palindrome_LL.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


"""
h       (none)
M -> A -> D <- A <- M      None
          s
                    f
                            curr 
                    prev
                            next

                        prev
M   ->  A   ->  D   ->  none

M   ->  A   ->  D   ->  none    
                        head                                                        
"""


def isPalindrome(head):
    # find the middle element
    slow = head
    fast = head

    while fast and fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    # reversing the linkedlist from middle element

    prev = None
    curr = slow

    while curr:
        nextNode = curr.next
        curr.next = prev
        prev = curr
        curr = nextNode

    # check head node and prev node values

    while head:
        if head.val != prev.val:
            return False
        head = head.next
        prev = prev.next

    return True


headNode = Node("M")
headNode.next = Node("M")
headNode.next.next = Node("A")
headNode.next.next.next = Node("D")
headNode.next.next.next.next = Node("A")
headNode.next.next.next.next.next = Node("M")
headNode.next.next.next.next.next.next = Node("M")
print(isPalindrome(headNode))


# TC: O(N)
# SC: O(1)

File: JavaScript/LinkedList/Scaler/palindrome_ll.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function isPalindrome(head) {
  // Find the middle element using the slow and fast pointers
  let slow = head;
  let fast = head;

  while (fast && fast.next && fast.next.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  // Reverse the linked list from the middle element
  let prev = null;
  let curr = slow;

  while (curr) {
    let nextNode = curr.next;
    curr.next = prev;
    prev = curr;
    curr = nextNode;
  }

  // Check head node and prev node values
  while (head) {
    if (head.val !== prev.val) {
      return false;
    }
    head = head.next;
    prev = prev.next;
  }

  return true;
}

// Create nodes for the linked list
let headNode = new Node("M");
headNode.next = new Node("M");
headNode.next.next = new Node("A");
headNode.next.next.next = new Node("D");
headNode.next.next.next.next = new Node("A");
headNode.next.next.next.next.next = new Node("M");
headNode.next.next.next.next.next.next = new Node("M");

console.log(isPalindrome(headNode));

// TC: O(N)
// SC: O(1)
